File: csharp/interest-is-interesting/InterestIsInteresting.cs
using System;

static class SavingsAccount
{
    public static float InterestRate(decimal balance)
    {
        var output = 0.0f;

        switch (balance){

            case < 0 : output = 3.213f;
                        break;
            case >= 0 when balance < 1000.0m : output =  0.5f;
                        break;
            case > 0 when balance >= 1000.0m && balance < 5000.0m : output = 1.621f;
                        break;
            case > 0 when balance >= 5000.0m : output = 2.475f;
                        break;
        }
        return output;
    }

    public static decimal Interest(decimal balance)
    {
        return (balance * (decimal)InterestRate(balance)/100.0m);
    }

    public static decimal AnnualBalanceUpdate(decimal balance)
    {
        var r = (decimal)InterestRate(balance)/100.0m;

        return balance*(1.0m +r*1.0m);  
    }

    public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
    {
        var years = 0;
        var actualBalance = balance;
        while (actualBalance < targetBalance && balance >= 0 || actualBalance > targetBalance && balance < 0){
            actualBalance = AnnualBalanceUpdate(actualBalance);
            years++;
        }
        return years;
    }
}

File: cpp/interest-is-interesting/interest_is_interesting.cpp
#include <iostream>

// interest_rate returns the interest rate for the provided balance.
double interest_rate(double balance) {
    if(balance >= 5000) return 2.475; 
    if(balance >= 1000) return 1.621;
    if(balance >= 0) return 0.5;

    return 3.213;
}

// yearly_interest calculates the yearly interest for the provided balance.
double yearly_interest(double balance) {
    return balance * (interest_rate(balance)/100);
}

// annual_balance_update calculates the annual balance update, taking into
// account the interest rate.
double annual_balance_update(double balance) {
    return balance + yearly_interest(balance);
}

// years_until_desired_balance calculates the minimum number of years required
// to reach the desired balance.
int years_until_desired_balance(double balance, double target_balance) {
    double new_balance = balance;
    int years = 0;

    for (int i = 1; new_balance < target_balance; i++){
        new_balance = annual_balance_update(new_balance);
        years = i;
    }

    return years;
}
